src/Smalot/PdfParser/Font.php

```diff
@@ -381,86 +393,182 @@ public function decodeText(array $commands): string
-            $result = '';
-            if ($unicode) {
-                $chars = preg_split(
-                        '//s'.($unicode ? 'u' : ''),
-                        $text,
-                        -1,
-                        \PREG_SPLIT_DELIM_CAPTURE | \PREG_SPLIT_NO_EMPTY
-                );
-
-                foreach ($chars as $char) {
-                    $dec_av = hexdec(bin2hex($char));
-                    $dec_ap = $encoding->translateChar($dec_av);
-                    $result .= self::uchr($dec_ap ?? $dec_av);
-                }
-            } else {
-                $length = \strlen($text);
 
-                for ($i = 0; $i < $length; ++$i) {
-                    $dec_av = hexdec(bin2hex($text[$i]));
-                    $dec_ap = $encoding->translateChar($dec_av);
-                    $result .= self::uchr($dec_ap ?? $dec_av);
-                }
-            }
             $text = $result;
-        } elseif ($this->get('Encoding') instanceof Element &&
-                  $this->get('Encoding')->equals('MacRomanEncoding')) {
-            // mb_convert_encoding does not support MacRoman/macintosh,
-            // so we use iconv() here
-            $text = iconv('macintosh', 'UTF-8', $text);
-        } elseif (!mb_check_encoding($text, 'UTF-8')) {
-            // don't double-encode strings already in UTF-8
-            $text = mb_convert_encoding($text, 'UTF-8', 'Windows-1252');
         }
 
         return $text;
     }
+
+    /**
+     * Decode content by any type of Encoding (dictionary's item) instance.
+     */
+    private function decodeContentByEncoding(string $text): ?string
+    {
+        $encoding = $this->get('Encoding');
+
+        // When Encoding referenced by object id but object itself not contains `/Type /Encoding` in it's dictionary.
+        if ($encoding instanceof PDFObject) {
+            $encoding = $this->getInitializedEncodingByPdfObject($encoding);
+        }
+
+        // When Encoding contains `/Type /Encoding` in it's dictionary.
+        if ($encoding instanceof Encoding) {
+            return $this->decodeContentByEncodingEncoding($text, $encoding);
+        }
+
+        // When Encoding is just String
+        if ($encoding instanceof Element) { //todo: ElementString class must by used?
+            return $this->decodeContentByEncodingElement($text, $encoding);
+        }
+
+        // Encoding has unintended type.
+        $encodingClassName = \get_class($encoding);
+        throw new LogicException("Unknown encoding instance type: {$encodingClassName}");
+    }
+
+    private function getInitializedEncodingByPdfObject(PDFObject $PDFObject): Encoding
+    {
+        if (!$this->initializedEncodingByPdfObject) {
+            $this->initializedEncodingByPdfObject = $this->createInitializedEncodingByPdfObject($PDFObject);
+        }
+
+        return $this->initializedEncodingByPdfObject;
+    }
+
+    /**
+     * Decode content when Encoding is instance of Encoding.
```

Comment:
This sentence does not make sense in my opinion. `$encoding` is always of type `Encoding`, isn't it?

Reply:
I assumed it, but the sentence is weird because you say X is and instance of X. But you refrain to `$this->get('Encoding')`, don't you?

My suggestion would be:

> Decode content when $this->get('Encoding') is instance of Encoding.

Would you agree to that?

Reply:
I know what you want to say but in the sentence you basically say "X is an instance of X". But you mean $this->get('Encoding') instead, don't you?

Just now I understand you))

> Decode content when $this->get('Encoding') is instance of Encoding.

Definitely, it would be much better than before!)

I used a quite changed version like:
```
Decode content when $encoding (given by $this->get('Encoding')) is instance of Encoding.
```

Reply:
> Decode content when $encoding (given by $this->get('Encoding')) is instance of Encoding.

That is a good description.